include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,192 @@
+//
+// Created by alandefreitas on 12/01/2022.
+//
+
+#ifndef BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+
+#include <atomic>
+
+#include <boost/url/detail/over_allocator.hpp>
+
+namespace boost {
+namespace urls {
+
+// base class to store the underlying allocator bound to `base`
+template <class T>
+struct any_allocator<T>::base
+{
+  std::atomic<std::size_t> refs{1};
+
+  virtual
+  ~base()
+  {
+  }
+
+  // although the allocator in `base` is bound to `base`, `allocate` always
+  // returns `char` as make_string_value does.
+  virtual
+  char*
+  allocate(std::size_t n) = 0;
+
+  virtual
+  void
+  deallocate(char*p, std::size_t n) = 0;
+
+  template <class U>
+  void
+  deallocate(U* p, std::size_t n) {
+      deallocate(reinterpret_cast<char*>(p), n * sizeof(T));
+  }
+
+  void
+  deallocate(char*p) {
+      deallocate(p, 1);
+  }
+
+  template <class U>
+  void
+  deallocate(U* p) {
+      deallocate(p, 1);
+  }
+
+  virtual void destroy_self() noexcept = 0;
```

Comment:
`destroy` has a different meaning in the named requirement.

Reply:
`destroy` is not public and it isn't part of a class which satisfies _Allocator_

Reply:
That's right. It doesn't satisfy Allocator because it's our limited allocator to erase the type. But it does partially match the allocator API when possible so we can easily forward calls from any_allocator to base, so I just thought `destroy` would be unnecessarily confusing.

Reply:
when you have a choice between more characters or less, you should always choose less

Reply:
This function hasn't existed since we moved to shared_ptr.